### PR TITLE
LogitNet weights

### PR DESCRIPTION
**Pull request by eli-goodfriend**
This PR allows weights to be used in fitting models with LogitNet.

The underlying Fortran glmnet code does not use weights in computing the log loss during gradient descent. To work around this limitation, the R glmnet wrapper approximates the effect of weighting by [multiplying the target by the weights](https://github.com/cran/glmnet/blob/master/R/lognet.R).  

While not exactly correct, this weighting can improve performance when fitting logistic regression models on datasets with class imbalance. This improvement is demonstrated in the included test.

### CODE REVIEW

File: glmnet/logistic.py [hunk 1]
```diff
@@ -266,6 +266,17 @@ def _fit(self, X, y, sample_weight=None, relative_penalties=None):
             # keep the original order.
             _y = (y[:, None] == self.classes_).astype(np.float64, order='F')
 
+        # use sample weights, making sure all weights are positive
+        # this is inspired by the R wrapper for glmnet, in lognet.R
+        if sample_weight is not None:
+            weight_gt_0 = sample_weight > 0
+            if not np.all(weight_gt_0):
```
Comment (xdavio) [inline glmnet/logistic.py]: The three lines after it should work when all weights are positive. I have a slight preference for avoiding conditionals when possible.

File: glmnet/logistic.py [hunk 2]
```diff
@@ -266,6 +266,17 @@ def _fit(self, X, y, sample_weight=None, relative_penalties=None):
...
+                sample_weight = sample_weight[weight_gt_0]
+                _y = _y[weight_gt_0, :]
+                X = X[weight_gt_0, :]
+            sample_weight = np.expand_dims(sample_weight, 1)
+            _y = _y * np.tile(sample_weight, (1, n_classes))
```
Comment (xdavio) [inline glmnet/logistic.py]: Numpy broadcasting would simplify these two lines a little bit

**Comment by xdavio** [comment]
Another thought: it might be worthwhile to also do the tests for when `X` is sparse, because this PR ends up subsetting `X` (which might be a sparse CSR matrix)

**Comment by eli-goodfriend** [comment]
Thanks! I switched the test to be with sparse X, I think that's a more rigorous test.

**Comment by xdavio** [comment]
[logitnet_weights_check.pdf](https://github.com/civisanalytics/python-glmnet/files/1737400/logitnet_weights_check.pdf)

Hey @eli-goodfriend , thanks for your patience on this. I think we're about ready to merge, but I wanted to put this pdf in front of you first.

I'm trying to validate this PR by computing the correlation between the residual and the weight magnitude. I expect the residual to decrease when the weight is large. The results seem to agree with this. 

If you're also convinced, drop a note here and we'll merge.

**Comment by eli-goodfriend** [comment]
Makes sense to me! Thanks for doing that validation.

**Comment by xdavio** [comment]
Thanks for the contribution. LGTM!
